packages/core/core/src/requests/EntryRequest.js

```diff
@@ -91,40 +92,70 @@ class EntryResolver {
     try {
       stat = await this.options.inputFS.stat(entry);
     } catch (err) {
-      throw new Error(`Entry ${entry} does not exist`);
+      throw new ThrowableDiagnostic({
+        diagnostic: {
+          message: `Entry ${entry} does not exist`,
+          filePath: entry,
+        },
+      });
     }
 
     if (stat.isDirectory()) {
       let pkg = await this.readPackage(entry);
-      if (pkg && typeof pkg.source === 'string') {
-        let source = path.join(path.dirname(pkg.filePath), pkg.source);
-        try {
-          stat = await this.options.inputFS.stat(source);
-        } catch (err) {
-          throw new Error(
-            `${pkg.source} in ${path.relative(
-              this.options.inputFS.cwd(),
-              pkg.filePath,
-            )}#source does not exist`,
-          );
+      if (pkg && pkg.source != null) {
+        let entries = [];
+        let files = [];
+
+        let pkgSources = Array.isArray(pkg.source) ? pkg.source : [pkg.source];
+        for (let pkgSource of pkgSources) {
+          if (typeof pkgSource === 'string') {
+            let source = path.join(path.dirname(pkg.filePath), pkgSource);
+            try {
+              stat = await this.options.inputFS.stat(source);
+            } catch (err) {
+              throw new ThrowableDiagnostic({
+                diagnostic: {
+                  message: `${pkgSource} in ${path.relative(
+                    this.options.inputFS.cwd(),
+                    pkg.filePath,
+                  )}#source does not exist`,
+                  filePath: source,
+                },
+              });
+            }
+
+            if (!stat.isFile()) {
+              throw new ThrowableDiagnostic({
+                diagnostic: {
+                  message: `${pkgSource} in ${path.relative(
+                    this.options.inputFS.cwd(),
+                    pkg.filePath,
+                  )}#source is not a file`,
+                  filePath: source,
+                },
+              });
+            }
+
+            entries.push({filePath: source, packagePath: entry});
+            files.push({filePath: pkg.filePath});
```

Comment:
@mischnic is this semantically correct? If pkg.source is an array then there will be the same { filePath: pkg.filePath } pushed for every source? seems like only a single one should be pushed.